### ISSUE TITLE
Gray out code excluded by preprocessor

### ISSUE THREAD

**Issue by grigorig** [clangd/vscode-clangd#149]
The official C++ extensions grays out code that is excluded by preprocessor definitions, which is very useful, particular in setups with many bits of configuration-dependent or architecture/OS-specific sections. It would be very useful if the clangd extension could do that as well.

**Comment by njames93** [comment]
This feature is supported in trunk and is on the next release branch, probably landing some time in March.

**Comment by HighCommander4** [comment]
Yes, this does work in clangd 12 using the new standard `semanticTokens` protocol. Each preprocessor-disabled line gets a `comment` semantic token spanning the text on the line. The appearance can be customized via `editor.semanticTokenColorCustomizations`.

There are a couple of limitations of this approach as compared to inactive code highlighting as seen in some other editors:

  1. The styling is limited to foreground color and text styling. You cannot, for instance, specify a grey **background** color for these lines (at least, not until https://github.com/microsoft/vscode/issues/3429 is resolved).
  2. Even if background styling is permitted, since the semantic token only covers the text on the line, only the portion of each inactive line that contains text will be colored, rather than the whole line, as in other editors.

I [wrote a patch](https://reviews.llvm.org/D67537) a while ago which implements a different approach, with a custom protocol between clangd and vscode-clangd, which allows styling the background of the whole line.

I'm not sure yet if there is a way to extend the current implementation to achieve this more flexible styling, or if I should resurrect my patch. Opinions are welcome.

**Comment by grigorig** [comment]
Sounds promising! I'm fine with just setting the foreground color to something distinctive, but if we can eventually something better, that's great.

**Comment by grigorig** [comment]
This actually seems to be a duplicate of #85.
